[mlir][bufferization] Empty tensor elimination based on SubsetOpInterface

matthias-springer opened the pull request:
This commit generalizes empty tensor elimination to operate on subset ops. No new test cases are added because all current subset ops were already supported previously. From this perspective, this change is NFC.

A new interface method (and a helper method) are added to `SubsetInsertionOpInterface` to build the subset of the destination tensor.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/TransformOps/BufferizationTransformOps.td:
```diff
@@ -109,19 +109,17 @@ def EliminateEmptyTensorsOp
-    `tensor.empty` ops cannot be bufferizes. They can either be converted to
-    `bufferization.alloc_tensor` or replaced with another tensor (via this
-    transform). `tensor.empty` does not specify the contents of the returned
+    "tensor.empty" ops cannot be bufferized. They can either be converted to
+    "bufferization.alloc_tensor" or replaced with another tensor (via this
+    transform). "tensor.empty" does not specify the contents of the returned
```
I think backticks are used when referring to function/variable names only.

maerhart replied:
Interesting, I often do the exact opposite. Is there a written guideline somewhere?

matthias-springer committed:
[mlir][bufferization] Empty tensor elimination based on SubsetOpInterface

This commit generalizes empty tensor elimination to operate on subset ops. No new test cases are added because all current subset ops were already supported by previously. From this perspective, this change is NFC.

A new interface method (and a helper method) are added to `SubsetOpInterface` to build the subset of the destination tensor.
